multiopython/plans/actions.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import annotations
+
+from typing import Any, Literal
+
+from pydantic import BaseModel, Field, ValidationInfo, WrapValidator, field_validator, validate_call
+from typing_extensions import Annotated
+
+from .sinks import SINKS, Sinks
+from .utils import make_validator
+
+
+def convert_to_sinks(v: Any) -> Sinks:
+    """
+    Convert a general dict to a Sinks instance
+    """
+    if not isinstance(v, (Sinks, dict)):
+        raise ValueError(f"Sinks must be a dict or an instance of Sinks, not {type(v)}")
+    if isinstance(v, Sinks):
+        return v
+    if "type" not in v:
+        raise ValueError("Sinks must have a type")
+    if v["type"] not in SINKS:
+        raise ValueError(f"Invalid sink type, must be one of {list(SINKS.keys())}")
+    sink = SINKS[v["type"]](**v)
+    return sink
+
+
+SinkType = Annotated[dict | Sinks, WrapValidator(make_validator(convert_to_sinks))]
+
```

Comment:
what you have works, but perhaps could be made more convenient and standard, if you use this https://docs.pydantic.dev/2.9/concepts/unions/#discriminated-unions

you already have a discriminator in place -- so you can just introduce another BaseModel, eg ActionHolder, which has a field `action: Select|Statistics|...`, and the discriminator. Then, the `convert_to_sinks` method and `ACTIONS` lookup won't be needed as you'd just initialize the ActionHolder instead

ftr, I generally don't like literals and prefer using enums -- not because of pydantic, but because I think they play better with mypy. But if you only need them for pydantic serde, its ok

Reply:
btw I'm getting a bit confused about why sinks parsing is in actions.py, and actions parsing is in plans.py -- is that on purpose?

Reply:
The location was a product of the prior way of validating it, thanks for the suggestion, I shall shift to that in the next commit